src/mstr.h

```diff
@@ -0,0 +1,223 @@
+ *    consideration when defining new struct in C. Note also that each metadata
+ *    might either be attached to mstr or not which complicates the design phase
+ *    of a new mstrKind a little.
+ *
+ *    In the example above, HKEY_META_VAL_REF_COUNT, with worst alignment of 4
+ *    bytes, is enumerated first, and therefore, will be last in memory layout.
+ *
+ * 2) Few optimizations in Redis rely on the fact that sds address is always an odd
+ *    pointer. We can achieve the same with a little effort. It was already taken
+ *    care that all headers of type mstrhdrX has odd size. With that in mind, if
+ *    a new kind of mstr is required to be limited to odd addresses, then we must
+ *    make sure that sizes of all related metadatas that are defined in mstrKind
+ *    are even in size.
+ */
+
+#ifndef __MSTR_H
+#define __MSTR_H
+
+#include <sys/types.h>
+#include <stdarg.h>
+#include <stdint.h>
+
+/* Selective copy of ifndef from server.h instead of including it */
+#ifndef static_assert
+#define static_assert(expr, lit) extern char __static_assert_failure[(expr) ? 1:-1]
+#endif
+
+#define MSTR_TYPE_5         0
+#define MSTR_TYPE_8         1
+#define MSTR_TYPE_16        2
+#define MSTR_TYPE_64        3
+#define MSTR_TYPE_MASK      3
+#define MSTR_TYPE_BITS      2
+
+#define MSTR_META_MASK      4
+
+#define MSTR_HDR(T,s) ((struct mstrhdr##T *)((s)-(sizeof(struct mstrhdr##T))))
+#define MSTR_HDR_VAR(T,s) struct mstrhdr##T *sh = (void*)((s)-(sizeof(struct mstrhdr##T)));
+
+#define MSTR_META_BITS  1  /* is metadata attached? */
+#define MSTR_TYPE_5_LEN(f) ((f) >> (MSTR_TYPE_BITS + MSTR_META_BITS))
+#define CREATE_MSTR_INFO(len, ismeta, type) ( (((len<<MSTR_META_BITS) + ismeta) << (MSTR_TYPE_BITS)) | type )
+
+/* mimic plain c-string */
+typedef char *mstr;
+
+/* Flags that can be set on mstring to indicate for attached metadata. It is
+ * */
+typedef uint16_t mstrFlags;
+
+struct __attribute__ ((__packed__)) mstrhdr5 {
+    unsigned char info; /* 2 lsb of type, 1 metadata, and 5 msb of string length */
+    char buf[];
+};
+struct __attribute__ ((__packed__)) mstrhdr8 {
+    uint8_t unused;  /* To achieve odd size header (See comment above) */
+    uint8_t len;
+    unsigned char info; /* 2 lsb of type, 6 unused bits */
+    char buf[];
+};
+struct __attribute__ ((__packed__)) mstrhdr16 {
+    uint16_t len;
+    unsigned char info; /* 2 lsb of type, 6 unused bits */
+    char buf[];
+};
+struct __attribute__ ((__packed__)) mstrhdr64 {
+    uint64_t len;
+    unsigned char info; /* 2 lsb of type, 6 unused bits */
+    char buf[];
+};
+
+#define NUM_MSTR_FLAGS (sizeof(mstrFlags)*8)
+
+/* mstrKind is used to define a kind (a group) of mstring with its own metadata layout */
+ typedef struct mstrKind {
+    const char *name;
+    int metaSize[NUM_MSTR_FLAGS];
+} mstrKind;
+
+mstr mstrNew(const char *initStr, size_t lenStr);
+
+mstr mstrNewWithMeta(struct mstrKind *kind, const char *initStr, size_t lenStr, mstrFlags flags);
```

Comment:
I'm wodering about the design of this api, why the metadata structure isn't determined by the called, but flags is used as it is not.
like:

```
mstrKind mstrFieldKind = {
        .name = "hField",

        /* Taking care that all metaSize[*] values are even ensures that all
         * addresses of hfield instances will be odd. */
        .metaSize = sizeof(ExpireMeta),
};

void *mstrMeta(mstr s, struct mstrKind *kind);
mstr mstrNewWithMeta(struct mstrKind *kind, const char *initStr, size_t lenStr);
```

Reply:
Quote from above:
> each hash-field instance still has the freedom to optionally attach the expire metadata to it. Most of them will start their life without any metadata attached.

That is, on instantiation, at mstr level, `mstrFlags flags` indicates for each given instance which metadata to allocate.